packages/typespec-client-generator-core/src/validate.ts

```diff
@@ -1,12 +1,160 @@
-import { Program } from "@typespec/compiler";
-import { createTCGCContext } from "./internal-utils.js";
-import { getAllModelsWithDiagnostics } from "./types.js";
+import {
+  AugmentDecoratorStatementNode,
+  DecoratorExpressionNode,
+  Enum,
+  EnumMember,
+  Interface,
+  Model,
+  ModelProperty,
+  Namespace,
+  Operation,
+  Program,
+  Scalar,
+  SyntaxKind,
+  Type,
+  Union,
+  UnionVariant,
+} from "@typespec/compiler";
+import { DuplicateTracker } from "@typespec/compiler/utils";
+import { getClientNameOverride } from "./decorators.js";
+import { AllScopes, clientNameKey, createTCGCContext, TCGCContext } from "./internal-utils.js";
+import { reportDiagnostic } from "./lib.js";
 
 export function $onValidate(program: Program) {
-  // Pass along any diagnostics that might be returned from the HTTP library
   const tcgcContext = createTCGCContext(program);
-  const [_, diagnostics] = getAllModelsWithDiagnostics(tcgcContext);
-  if (diagnostics.length > 0) {
-    program.reportDiagnostics(diagnostics);
+  const languageScopes = getDefinedLanguageScopes(program);
+  for (const scope of languageScopes) {
+    validateClientNamesPerNamespace(program, tcgcContext, scope, program.getGlobalNamespaceType());
+  }
+}
+
+function getDefinedLanguageScopes(program: Program): Set<string | typeof AllScopes> {
+  const languageScopes = new Set<string | typeof AllScopes>();
+  for (const value of program.stateMap(clientNameKey).values()) {
+    if (value[AllScopes]) {
+      languageScopes.add(AllScopes);
+    }
+    for (const languageScope of Object.keys(value)) {
+      languageScopes.add(languageScope);
+    }
+  }
+  return languageScopes;
+}
+
+function validateClientNamesPerNamespace(
+  program: Program,
+  tcgcContext: TCGCContext,
```

Comment:
duplication, you could get program from context.

Reply:
good point, removed the duplicate